### PR TITLE
Install windows binary releases

### PR DESCRIPTION
**Pull request by genotrance**
Much cleaner than before

### CODE REVIEW

File: src/choosenim.nim [hunk 1]
```diff
@@ -30,13 +30,24 @@ proc parseVersion(versionStr: string): Version =
...
+
   # Install the requested version.
   let path = download(version, params)
   # Extract the downloaded file.
   let extractDir = params.getInstallationDir(version)
   # Make sure no stale files from previous installation exist.
   removeDir(extractDir)
-  extract(path, extractDir)
+  if path.splitFile().ext == ".zip":
+    extract(path, extractDir.parentDir())
```
Comment (dom96) [inline src/choosenim.nim]: `extract` shouldn't have different behaviour based on the archive you're extracting, this if branch should be in the `extract`.

File: src/choosenim.nim [hunk 2]
```diff
@@ -30,13 +30,24 @@ proc parseVersion(versionStr: string): Version =
   result = newVersion(versionStr)
 
 proc installVersion(version: Version, params: CliParams) =
+  when defined(windows):
+    # Add MingW bin dir to PATH so download step can find gcc.
+    if findExe("gcc") == "" and dirExists(params.getMingwBin()):
+      let pathEnv = getEnv("PATH")
+      putEnv("PATH", params.getMingwBin() & PathSep & pathEnv)
+      defer:
+        putEnv("PATH", pathEnv)
```
Comment (dom96) [inline src/choosenim.nim]: Please move this to the appropriate place in ``download`` (it should actually go into ``getGccArch`` I think?).

Comment (dom96) [reply src/choosenim.nim]: Even if it means code duplication in ``build``.

File: src/choosenim/download.nim [hunk 3]
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
...
+      url = winBinaryZipUrl % [$version, $arch]
+      if not needsDownload(params, url, outputPath): return outputPath
+      try:
+        downloadFile(url, outputPath, params)
+        return outputPath
+      except HttpRequestError:
+        display("Info:", "Binary build unavailable, building from source",
+                priority = HighPriority)
+
+    url = websiteUrl % $version
```
Comment (dom96) [inline src/choosenim/download.nim]:
```suggestion
```

File: src/choosenim/download.nim [hunk 4]
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
   else:
     display("Downloading", "Nim $1 from $2" % [$version, "nim-lang.org"],
             priority = HighPriority)
-    let url = websiteUrl % $version
-    var outputPath: string
+
+    var
+      url: string
+      outputPath: string
```
Comment (dom96) [inline src/choosenim/download.nim]: Move these definitions below the `when`

File: src/choosenim/download.nim [hunk 5]
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
...
+    when defined(Windows):
+      url = winBinaryZipUrl % [$version, $arch]
```
Comment (dom96) [inline src/choosenim/download.nim]:
```suggestion
      let winUrl = winBinaryZipUrl % [$version, $arch]
```

File: src/choosenim/download.nim [hunk 6]
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
...
+      if not needsDownload(params, url, outputPath): return outputPath
```
Comment (dom96) [inline src/choosenim/download.nim]:
```suggestion
      if not needsDownload(params, winUrl, outputPath): return outputPath
```

File: src/choosenim/download.nim [hunk 7]
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
...
+      try:
+        downloadFile(url, outputPath, params)
```
Comment (dom96) [inline src/choosenim/download.nim]:
```suggestion
        downloadFile(winUrl, outputPath, params)
```

File: src/choosenim/download.nim [hunk 8]
```diff
@@ -217,8 +219,21 @@ proc downloadImpl(version: Version, params: CliParams): string =
   else:
     display("Downloading", "Nim $1 from $2" % [$version, "nim-lang.org"],
             priority = HighPriority)
-    let url = websiteUrl % $version
-    var outputPath: string
+
+    var
+      url: string
+      outputPath: string
+    when defined(Windows):
```
Comment (dom96) [inline src/choosenim/download.nim]: It seems like it would be relatively easy to check for the presence of `powershell` on the system so that we can continue to support Windows XP. Please do this by either:

* Using the ``osinfo`` package
* Running `powershell -v` (guessing the syntax here, just something that will always work on systems with powershell) and verifying it succeeds

If powershell isn't supported then we want to build from source.

File: src/choosenim.nim (hunk 1 above)
Comment (genotrance) [reply src/choosenim.nim]: This is only because the zip file contents already has a `nim-$version` folder in it. Same will apply to xz files as well when we migrate to nimarchive. I can move it to `extract()` if preferred though.

**Comment by genotrance** [comment]
I have addressed all your concerns - PTAL.

BTW, CI ran fine on my [instance](https://travis-ci.org/genotrance/choosenim/builds/479256174) of Travis but failed here for some reason.

**Comment by genotrance** [comment]
Okay passed this time - looks like tester has some issue in detecting successful pass.

File: src/choosenim/switcher.nim [hunk 9]
```diff
@@ -42,8 +42,12 @@ proc areProxiesInstalled(params: CliParams, proxies: openarray[string]): bool =
     if contents != proxyExe:
       return false
 
-proc isCCInPath(params: CliParams): bool =
-  return findExe("gcc") != "" or findExe("clang") != ""
+proc isCCInPath*(params: CliParams): bool =
+  when defined(Windows):
+    return findExe("gcc") != ""
+  else:
+    # Fix issue #104
```
Comment (dom96) [inline src/choosenim/switcher.nim]: This statement is ambiguous, is it a TODO or a "This code fixes issue #104"?

I'm assuming the former, but you seem to already be sort of fixing it? Just fix it properly.

* For Windows: check for gcc only
* For macOS (or freebsd): check for clang only
* For Linux: check for gcc only

And maybe rename this function to `isDefaultCCInPath` since these are the default C compilers used by Nim on these platforms.

File: src/choosenim/builder.nim [hunk 10]
```diff
@@ -82,7 +82,9 @@ proc build*(extractDir: string, version: Version, params: CliParams) =
   setCurrentDir(extractDir)
   # Add MingW bin dir to PATH so that `build.bat` script can find gcc.
   let pathEnv = getEnv("PATH")
-  putEnv("PATH", params.getMingwBin() & PathSep & pathEnv)
+  when defined(windows):
+    if not isCCInPath(params) and dirExists(params.getMingwBin()):
```
Comment (dom96) [inline src/choosenim/builder.nim]: Why is this necessary?

File: src/choosenim/download.nim [hunk 11]
```diff
@@ -199,6 +200,15 @@ proc needsDownload(params: CliParams, downloadUrl: string,
     return false
 
 proc downloadImpl(version: Version, params: CliParams): string =
+  # Add MingW bin dir to PATH so getGccArch can find gcc.
```
Comment (dom96) [inline src/choosenim/download.nim]: without this ``getGccArch`` won't work, so this code should be inside ``getGccArch``

File: src/choosenim.nim [hunk 12]
```diff
@@ -36,7 +36,12 @@ proc installVersion(version: Version, params: CliParams) =
   let extractDir = params.getInstallationDir(version)
   # Make sure no stale files from previous installation exist.
   removeDir(extractDir)
-  extract(path, extractDir)
+  if ("nim-$1_x" % $version) in path.splitFile().name:
```
Comment (dom96) [inline src/choosenim.nim]: Pretty sure this will break installation on other platforms. Also, my ``untar`` library already does this, so your new proc ``extractZip`` should do the same.

See how I do it in ``untar`` and do the same thing: https://github.com/dom96/untar/blob/master/src/untar.nim#L127 (also note that ``-j`` can be used for ``unzip``)

File: src/choosenim/switcher.nim (hunk 9 above)
Comment (genotrance) [reply src/choosenim/switcher.nim]: I was fixing it based on our IRC conversation. I've updated it.

File: src/choosenim/builder.nim (hunk 10 above)
Comment (genotrance) [reply src/choosenim/builder.nim]: It is consistent with the other instance. And it is only needed on Windows so this makes it clear.

File: src/choosenim/download.nim (hunk 11 above)
Comment (genotrance) [reply src/choosenim/download.nim]: I put it in `downloadImpl()` to avoid having to propagate `params` all over the place and minimize changes. I've changed it again.

File: src/choosenim.nim (hunk 12 above)
Comment (genotrance) [reply src/choosenim.nim]: Note that `-j` will get rid of all directory structure, not just the base dir. It will extract all files into the destination as one big blob.

I'll do this in extractZip().
